contracts/src/CrossChainApplications/NativeTokenBridge/tests/NativeTokenDestinationTests.t.sol

```diff
@@ -0,0 +1,396 @@
+            _DEFAULT_OTHER_BRIDGE_ADDRESS,
+            abi.encode(_DEFAULT_RECIPIENT, _DEFAULT_TRANSFER_AMOUNT)
+        );
+    }
+
+    function testInvalidSourceBlockchain() public {
+        vm.expectRevert(
+            _formatNativeTokenDestinationErrorMessage("invalid source chain")
+        );
+
+        vm.prank(MOCK_TELEPORTER_MESSENGER_ADDRESS);
+        nativeTokenDestination.receiveTeleporterMessage(
+            _MOCK_BLOCKCHAIN_ID,
+            _DEFAULT_OTHER_BRIDGE_ADDRESS,
+            abi.encode(_DEFAULT_RECIPIENT, _DEFAULT_TRANSFER_AMOUNT)
+        );
+    }
+
+    function testInvalidSenderContract() public {
+        vm.expectRevert(
+            _formatNativeTokenDestinationErrorMessage("unauthorized sender")
+        );
+
+        vm.prank(MOCK_TELEPORTER_MESSENGER_ADDRESS);
+        nativeTokenDestination.receiveTeleporterMessage(
+            _DEFAULT_OTHER_CHAIN_ID,
+            address(0x123),
+            abi.encode(_DEFAULT_RECIPIENT, _DEFAULT_TRANSFER_AMOUNT)
+        );
+    }
+
+    function testInvalidRecipientAddress() public {
+        vm.expectRevert(
+            _formatNativeTokenDestinationErrorMessage("zero recipient address")
+        );
+
+        vm.prank(MOCK_TELEPORTER_MESSENGER_ADDRESS);
+        nativeTokenDestination.receiveTeleporterMessage(
+            _DEFAULT_OTHER_CHAIN_ID,
+            _DEFAULT_OTHER_BRIDGE_ADDRESS,
+            abi.encode(address(0x0), _DEFAULT_TRANSFER_AMOUNT)
+        );
+    }
+
+    function testInvalidTransferAmount() public {
+        vm.expectRevert(
+            _formatNativeTokenDestinationErrorMessage("zero transfer value")
+        );
+
+        vm.prank(MOCK_TELEPORTER_MESSENGER_ADDRESS);
+        nativeTokenDestination.receiveTeleporterMessage(
+            _DEFAULT_OTHER_CHAIN_ID,
+            _DEFAULT_OTHER_BRIDGE_ADDRESS,
+            abi.encode(_DEFAULT_RECIPIENT, 0)
+        );
+    }
+
+    function testZeroRecipient() public {
+        collateralizeBridge();
+        vm.expectRevert(
+            _formatNativeTokenDestinationErrorMessage("zero recipient address")
+        );
+
+        nativeTokenDestination.transferToSource{
+            value: _DEFAULT_TRANSFER_AMOUNT
+        }(
+            address(0x0),
+            TeleporterFeeInfo({
+                feeTokenAddress: address(mockERC20),
+                amount: _DEFAULT_FEE_AMOUNT
+            }),
+            new address[](0)
+        );
+    }
+
+    function testUncollateralizedBridge() public {
+        vm.expectRevert(
+            _formatNativeTokenDestinationErrorMessage(
+                "contract undercollateralized"
+            )
+        );
+
+        nativeTokenDestination.transferToSource{
+            value: _DEFAULT_TRANSFER_AMOUNT
+        }(
+            _DEFAULT_RECIPIENT,
+            TeleporterFeeInfo({
+                feeTokenAddress: address(mockERC20),
+                amount: _DEFAULT_FEE_AMOUNT
+            }),
+            new address[](0)
+        );
+    }
+
```

Comment:
Can you add a test case for:
1. Adding less collateral than the token imbalance, and checking that the imbalance is properly decreased but supply left unchanged.
2. Adding more collateral than the token imbalance, and checking that the imbalance is decreased to 0 and supply incremented by the difference?